### DISCUSSION TITLE
Support `this params T[]`

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5659] [Language Ideas] [answered]
Supposed you have a bunch of variables and you want to calculate something with them, sometimes it's 4 sometimes it's 5, so you can write a function :
```
public static T Calculate (T[] array_);
```
but that means you have to construct an array and pass it in everytimes, good news it's already possible to use the ```params``` keyword for automating this process :
```
public static T Calculate  (params T[] array_);
//usage
MyStaticClass.Calculate(var1, var2, var3);
```
but what if you don't want to type the static class name and the method name? good news there exist something called an extension method that allows you to do it, bad news is it doesn't work with the params keyword :
```
//Extension method
public static T Calculate (this params T[] array_); //unsupported
//usage
(var1,var2,var3).Calculate();
```
If you would like this to be supported, you can leave a thumps up in the post, and if you are interested in this feature, please remember to comment and subscribe, to get the latest updates about this topic.

**Comment by 333fred** [comment]
I've edited the discussion title to remove the clickbait headline. Please try to avoid such in the future.

**Comment by YairHalberstadt** [comment]
> `(var1,var2,var3).Calculate();`

This already had s meaning - it's calling an extension method on a tuple.

> If you would like this to be supported, you can leave a thumps up in the post, and if you are interested in this feature, please remember to comment and subscribe, to get the latest updates about this topic.

This is the case for every discussion. Putting it here is just annoying.

**Comment by huoyaoyuan** [comment]
I don't think it's useful at all. Meanwhile, I don't think it's harmful either. The two features can be enabled orthogonally.

**Comment by HaloFour** [comment, marked as answer]
Assuming [collection literals](https://github.com/dotnet/csharplang/issues/5354) have a natural type (likely array) and permit dotting off of them directly you'll likely be able to write this without the need of a special `this params`.

```cs
[var1,var2,var3].Calculate();
```

Or you could accomplish this today by defining extension methods on tuples up to a certain arity:

```cs
using System;

public static class MyStaticClass {
    public static int Calculate(params int[] values) => 0;
    
    public static int Calculate(this (int, int) values) {
        return Calculate(values.Item1, values.Item2);
    }
    
    public static int Calculate(this (int, int, int) values) {
        return Calculate(values.Item1, values.Item2, values.Item3);
    }
    
    public static int Calculate(this (int, int, int, int) values) {
        return Calculate(values.Item1, values.Item2, values.Item3, values.Item4);
    }
}


public class C {
    public void M() {
        var result = (1, 2, 3).Calculate();
    }
}
```

**Reply by CyrusNajmabadi** [reply]
> you'll likely be able to write this without the need of a special this params.

Agreed. This sort of case also motivates me to have these be `Span<T>` so that you don't have any allocation penalty here. We will def be discussing this.

**Reply by Xyncgas** [reply]
> Or you could accomplish this today by defining extension methods on tuples up to a certain arity:

Maybe there should be a way to write generic with variable number of type params, there can be some constrains say maybe they are all the same type, then we can do so ```static TResult Calculate<params T> (this T);```

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/discussions/2330

Given generics are not expanded at compile time like templates it's hard to envision how support for variadic generics could be added to .NET.

**Reply by Xyncgas** [reply]
When I came to c# from c++ it's one of the first thing I noticed. I used to be writing templates in c++ and there seems to not be a way to do what I was used to be doing, still generics in c# is quite useful. This was then, now we have source generator
